### ISSUE TITLE
[HF transformers] LayerDrop RNG diverges under torch.compile(backend="eager") for audio encoders in train mode

### ISSUE THREAD

**Issue by penguinwu** [penguinwu/oss-model-graph-break-corpus#130]
## Summary

Audio encoder models (Wav2Vec2, Hubert, UniSpeech, UniSpeechSat, Data2VecAudio, WavLM, SEW, Wav2Vec2Conformer) produce **drastically different output (max abs diff ~5.0)** between eager and `torch.compile(model, backend="eager", fullgraph=False)` when the model is in **train mode** (`model.train()`).

Root cause: the encoder's per-layer **LayerDrop** decision is a stochastic `torch.rand([])` draw. Under `torch.compile` with graph breaks (the default, `fullgraph=False`), the RNG state seen by the compiled regions diverges from eager, so a **different set of layers is dropped** — changing the output by O(1), not just at the numerical-noise level.

This is the companion to the SDPA `_ignore_causal_mask_sdpa` issue; together the two account for the bulk of the bitwise-equivalence gap for HF models under `backend="eager"`.

## Environment

- Sweep / reproduction ran on transformers 5.6.2, but the LayerDrop loop is **unchanged in every release through 5.9.0 (the current latest, released 2026-05-20)** — verified directly against the v5.9.0 source: `modeling_wav2vec2.py` L701-703 (and the analogous loops in the other audio-encoder model files). The fix applies cleanly to current `main`.
- torch 2.13 nightly (cu126); reproduces on stable torch ≥ 2.3
- CUDA GPU, `model.train()`

## Reproduction

```python
import torch, torch._dynamo
from transformers import Wav2Vec2Model, Wav2Vec2Config

torch.manual_seed(0)
model = Wav2Vec2Model(Wav2Vec2Config()).cuda().train()
x = torch.randn(1, 16000).cuda()

torch._dynamo.reset(); torch.manual_seed(0)
out_eager = model(x).last_hidden_state

torch._dynamo.reset(); torch.manual_seed(0)
compiled = torch.compile(model, backend="eager", fullgraph=False)
out_compile = compiled(x).last_hidden_state

print("max abs diff:", (out_eager - out_compile).abs().max().item())  # ~5.0
```

## Root cause

`transformers/models/wav2vec2/modeling_wav2vec2.py` (and the analogous encoder loops in `hubert`, `unispeech`, `unispeech_sat`, `data2vec_audio`, `wavlm`, `sew`, `wav2vec2_conformer`):

```python
for layer in self.layers:
    dropout_probability = torch.rand([])
    skip_the_layer = self.training and dropout_probability < self.config.layerdrop
    if not skip_the_layer or synced_gpus:
        layer_outputs = layer(...)
    if skip_the_layer:
        layer_outputs = (None, None)
```

`torch.rand([])` is evaluated differently under compiled vs eager execution (graph-break re-entry advances Dynamo-captured RNG state out of step with eager), so `skip_the_layer` differs between the two runs → different layers are dropped → O(1) output divergence.

Ordinary `nn.Dropout` inside the layers contributes a smaller, additive divergence on top; for the models above, eliminating the LayerDrop/Dropout draws is what removes the large O(1) gap (see evidence).

## Evidence

Setting all dropout/LayerDrop probabilities to 0 reduces divergence from ~5.0 to ~0.14 (~35×). For **SEW, WavLM, and Wav2Vec2Conformer**, disabling LayerDrop/Dropout *alone* yields bitwise-equal output. For the other five, the residual ~0.14 is the separate `_ignore_causal_mask_sdpa` SDPA issue (companion); disabling LayerDrop **and** applying the SDPA fix yields bitwise-equal output for all 8 models.

## Proposed fix (HF-side)

Skip the stochastic LayerDrop draw under `torch.compile`, keeping it for ordinary eager training:

```python
from transformers.utils.import_utils import is_torchdynamo_compiling

for layer in self.layers:
    if self.training and not is_torchdynamo_compiling():
        skip_the_layer = torch.rand([]) < self.config.layerdrop
    else:
        skip_the_layer = False   # deterministic under torch.compile
    if not skip_the_layer or synced_gpus:
        layer_outputs = layer(...)
    if skip_the_layer:
        layer_outputs = (None, None)
```

This trades stochastic LayerDrop regularization for bitwise determinism under `torch.compile` — a reasonable tradeoff, since users opting into `torch.compile` for training are generally optimizing throughput, and LayerDrop under a no-op backend is already not behaving as intended (its random draws do not match eager). Apply consistently to all the audio-encoder model files listed above.

## Validation

1. Run the reproduction; after the patch assert `torch.equal(out_eager, out_compile)` is `True` for SEW / WavLM / Wav2Vec2Conformer, and (combined with the SDPA fix) for Wav2Vec2 / Hubert / UniSpeech / UniSpeechSat / Data2VecAudio.
2. Confirm eager-only training is unchanged (LayerDrop still active when not compiling).

## Longer-term / upstream note

The deeper fix is for `torch.compile` to advance RNG state across graph-break re-entry equivalently to eager. That is a PyTorch-side determinism concern; the HF-side patch above is the pragmatic near-term fix.
